src/smd-core.c

```diff
@@ -146,7 +146,7 @@ static int smd_attr_copy_val_to_internal(char *out, smd_dtype_t *t, smd_dtype_t
         case (SMD_TYPE_INT16): {
           int16_t ov = *(int16_t*)val;
           // check the range
-          if(ov < SCHAR_MIN || ov > SCHAR_MAX){
+          if(ov <= INT8_MIN || ov >= INT8_MAX){
```

Comment:
The equal sign here is wrong as `INT8_MIN` is indeed the smallest *representable* value (-128 in this case). You do not want to exclude the two legal values `INT8_MIN` and `INT8_MAX`, you only want to exclude the *non-representable* values `ov < -128 || ov > 127`.

The same holds true for all the other comparisons in this commit.

Reply:
Let's talk about this issue now. 

I just pushed one test called tst-float-int32.c with an example of the behaviour I'll describe here.

The goal: convert float f = INT32_MAX to int32.
The result: i32 = -INT32_MAX!!!

INT32_MAX = 2147483648

However, even smaller numbers, as 2147483600, are also converted to -2147483648. Problem after the 8-th digit, ok, but the problem shouldn't happen in the first place! The number 2147480000 is rightly converted.

To avoid this situation, I inserted the "equal" in the comparison (ov >= INT32_MAX) to prevent this wrong conversion which is not detected later since both stored numbers are equal and negative!

Then, for coherence (not tested!) I inserted the equal in all the comparisons and put a note indicating something was strange. There are more strange things happening and they were all reported to Julian. But he's still on vacation, so I'm not expecting an answer until next week.

A similar problem happens with uint32 which is even weirder because UINT32MAX = 2 x INT32MAX, so INT32MAX is way far from the limit. This problem is solved just changing the printf to %u, but this might be cheating. Inside gdb, however, the number is being saved as positive. In short, it's a mess to my simple mind.